OsmAnd/src/net/osmand/plus/plugins/osmedit/fragments/AdvancedEditPoiFragment.java

```diff
@@ -374,4 +381,51 @@ public static void addPoiToStringSet(AbstractPoiType abstractPoiType, Set<String
 					+ abstractPoiType.getClass());
 		}
 	}
+
+	private class OsmTagsArrayAdapter extends ArrayAdapter implements Filterable {
+
+		private OsmTagsFilter filter;
+
+		public OsmTagsArrayAdapter(Context context, int resource) {
+			super(context, resource);
+		}
+
+		@Override
+		public Filter getFilter() {
+			if (filter == null) {
+				filter = new OsmTagsFilter();
+			}
+			return filter;
+		}
+	}
+
+	private class OsmTagsFilter extends Filter {
+
+		@Override
+		protected FilterResults performFiltering(CharSequence constraint) {
+			FilterResults results = new FilterResults();
+			List<String> filteredTags = new ArrayList<String>();
+			String query = constraint.toString().trim();
+			for (String tag : allTags) {
+				if (tag.startsWith(query) || tag.contains(":" + query)) {
+					filteredTags.add(tag);
+				}
+			}
+			results.values = filteredTags;
+			results.count = filteredTags.size();
+			return results;
+		}
+
+		@SuppressWarnings("unchecked")
+		@Override
+		protected void publishResults(CharSequence constraint, FilterResults results) {
+			if (results.values != null) {
+				synchronized (mAdapter) {
```

Comment:
publishResults runs in UI Thread probably it's not neded

Reply:
Moreover `mAdapter` is not filter's adapter